### ISSUE TITLE
Questions related to real robot training

### ISSUE THREAD

**Issue by spomha** [cliport/cliport#22]
Hey Mohit, thank you for uploading such a detailed repository related to your work. I am trying to train a multi-task model on real robot using your code. I have few questions related to real robot setup and training and I'd appreciate if you could answer them when you're available:

1. How did you calculate the bounds, pix_size defined in dataset.py and why are these needed? 
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/dataset.py#L41-L44
My understanding with pix_size at least is that pix_size=1/max(self.in_shape[0], self.in_shape[1]). And you have hard coded this because your network has fixed input image size.

2. In what frame of reference are you saving your input demonstration action pairs position (XYZ)? For example, I initially thought you are saving them in camera coordinate system. But seeing the xyz_to_pix() conversion function in utils.py, you are not using camera intrinsic parameters to back project to pixel coordinate system.
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/utils/utils.py#L137-L141

3. What is the point of storing reward in input demonstrations? It seems like reward is always [0, 1], which is equally divided into time steps for a task. Also, I couldn't find it being used for updating loss functions i.e. attention_criterion/transport_criterion in transporter.py:
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/agents/transporter.py#L137-L175

3. Why is the p0_theta always set to 0 degrees in dataset.py and why do we set p1_theta to be a difference of p1 and p0 theta? Shouldn't we have different pick rotation angle similar to place rotation depending upon the object's orientation?
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/dataset.py#L185-L186

4. In your code, you are assuming the observations are taken from different views and fusing them i.e.:
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/utils/utils.py#L392-L406
In real robot with a single camera, do I need to consider any other changes in the code or will it be fine if I just ignore the views fusion step in get_fused_heightmap()

5. Where is 'zrange' being used from camera config defined in cameras.py? I couldn't pinpoint its usage in code:
https://github.com/cliport/cliport/blob/47bf8298f6b664365ab042eb7ac0bcaf5df60254/cliport/tasks/cameras.py#L105

Thank you for taking your time on this.

**Comment by MohitShridhar** [comment]
@spomha, happy to help!

1. I think `1/320 = 0.003125`. Yes, it's hardcoded because it's copied over code from TransporterNets

2. If I remember correctly, for real-robot experiments, everything was saved with respect to the robot base frame: the point clouds, pick and place poses. I think `xyz_to_pix` was only used for Ravens. For real poses from Franka, I had a custom function that did the 3D to 2D projection (with some basic camera projection algebra).

3. Good catch. There isn't. You don't need to save rewards, unless you plan to do some reinforcement learning. 

4. `p0_theta` is always zero because for suction-grippers, pick rotations don't matter since the contact is a point. However for parallel-jaw gripper you need a separate module that also predicts pick rotations. And yes, the `p1_theta` are delta predictions. This is a common thing in behavior-cloning where you predict delta-poses instead of absolute ones. 

5. You should be able to use the same code for real-obs fusion. You just need the camera parameters. This was my Kinect Config:
```
class KinectFranka():
    """Kinect Franka Camera"""

    # Near-orthographic projection.
    image_size = (424,512)
    intrinsics = (365.57489013671875, 0.0, 257.5205078125, 0.0, 365.57489013671875, 205.26710510253906, 0.0, 0.0, 1.0)
    position = (1.082, -0.041, 1.027)
    rotation = p.getQuaternionFromEuler((-2.611, 0.010, 1.553))

    # Camera config.
    CONFIG = [{
        'image_size': image_size,
        'intrinsics': intrinsics,
        'position': position,
        'rotation': rotation,
        'zrange': (0.01, 10.),
        'noise': False
    }]
 ```
 
 6. Oh hmmm... I think it's used [somewhere](https://github.com/cliport/cliport/blob/e9cde74754606448d8a0495c1efea36c29b201f1/cliport/environments/environment.py#L271). You can get them from your camera, or manually adjust them through visualization.

Hope this helps!

**Comment by spomha** [comment]
Hey Mohit, thank you for your swift response.

Regarding 4, could you briefly describe how should I go about implementing changes in pick module to handle discrete pick rotations? I am also using Franka robot in my workspace. To my understanding, I should follow similar logic given in place module i.e. :
https://github.com/cliport/cliport/blob/e9cde74754606448d8a0495c1efea36c29b201f1/cliport/models/streams/two_stream_transport_lang_fusion.py#L30-L67

**Comment by MohitShridhar** [comment]
@spomha, in addition to the pick module that predicts pick locations, I had a separate module for predicting rotations. Something like:

```
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F

from dvnets.utils import utils
from dvnets.models.core.two_stream_attention_lang_fusion import TwoStreamAttentionLangFusion
import dvnets.models as models
import dvnets.models.core.fusion as fusion


class TwoStreamAttentionRotationLangFusionLat(TwoStreamAttentionLangFusion):
    """Attention module."""

    def __init__(self, stream_backbone, in_shape, n_rotations, preprocess, cfg):
        self.fusion_type = cfg['train']['attn_stream_fusion_type']
        super().__init__(stream_backbone, in_shape, n_rotations, preprocess, cfg)

        self.crop_size = 64
        self.pad_size = int(self.crop_size / 2)
        self.padding = np.zeros((3, 2), dtype=int)
        self.padding[:2, :] = self.pad_size

    def _build_nets(self):
        # Initialize fully convolutional Residual Network with 43 layers and
        # 8-stride (3 2x2 max pools and 3 2x bilinear upsampling)
        stream_one_backbone, stream_two_backbone = self.stream_backbone
        stream_one_model = models.names[stream_one_backbone]
        stream_two_model = models.names[stream_two_backbone]

        in_shape = (64, 64, 6)
        self.attn_stream_one = stream_one_model(in_shape, 1, self.cfg)
        self.attn_stream_two = stream_two_model(in_shape, 1, self.cfg)
        self.fusion = fusion.names[self.fusion_type](input_dim=1)

        print(f"Attn Rotation Stream One: {stream_one_backbone}, Stream Two: {stream_two_backbone}")

    def attend(self, x, l):
        x1, lat = self.attn_stream_one(x)
        x2 = self.attn_stream_two(x, lat, l)
        x = self.fusion(x1, x2)
        return x

    def forward(self, in_img, p, lang_goal, softmax=True):
        """Forward pass."""
        in_data = np.pad(in_img, self.padding, mode='constant')
        # in_data = self.preprocess(in_data)
        in_shape = (1,) + in_data.shape
        in_data = in_data.reshape(in_shape)
        in_tens = torch.from_numpy(in_data).to(dtype=torch.float).cuda()

        # Pivot
        # pv = np.array(in_data.shape[1:3]) // 2
        pv = np.array([p[0], p[1]]) + self.pad_size

        # crop
        hcrop = self.crop_size//2
        in_tens = in_tens[:, pv[0]-hcrop:pv[0]+hcrop, pv[1]-hcrop:pv[1]+hcrop, :]

        # Rotate input.
        center = (hcrop, hcrop)
        in_tens = in_tens.permute(0, 3, 1, 2)
        in_tens = in_tens.repeat(self.n_rotations, 1, 1, 1)
        in_tens = self.rotator(in_tens, pivot=center)

        # Forward pass.
        logits = []
        for x in in_tens:
            logits.append(self.attend(x, lang_goal))
        logits = torch.cat(logits, dim=0)

        # Rotate back output.
        logits = self.rotator(logits, reverse=True, pivot=pv)
        logits = torch.cat(logits, dim=0)
        # c0 = self.padding[:2, 0]
        # c1 = c0 + in_img.shape[:2]
        # logits = logits[:, :, c0[0]:c1[0], c0[1]:c1[1]]

        logits = logits.permute(1, 2, 3, 0)  # D H W C
        output = logits.reshape(1, np.prod(logits.shape))
        if softmax:
            output = F.softmax(output, dim=-1)
            output = output.reshape(logits.shape[1:])
        return output
```

This module takes in the location predicted by the pick module, and is trained in a similar manner to place rotations.
